src/java/org/apache/cassandra/schema/MigrationCoordinator.java

```diff
@@ -355,11 +356,11 @@ private synchronized void removeEndpointFromVersion(InetAddressAndPort endpoint,
         FutureTask<Void> task = new FutureTask<>(() -> pullSchema(new Callback(endpoint, info)), null);
         if (shouldPullImmediately(endpoint, info.version))
         {
-            Stage.MIGRATION.submit(task);
+            MIGRATION.submit(task);
         }
         else
         {
-            ScheduledExecutors.nonPeriodicTasks.schedule(() -> Stage.MIGRATION.submit(task), MIGRATION_DELAY_IN_MS, TimeUnit.MILLISECONDS);
+            ScheduledExecutors.nonPeriodicTasks.schedule(() -> MIGRATION.submit(task), MIGRATION_DELAY_IN_MS, TimeUnit.MILLISECONDS);
```

Comment:
I think we need to do the shutdown check here, not after we've already submitted in `pullSchema()`

Reply:
I have one more suggestion, is it reasonable to add:
```
if (Stage.MIGRATION.executor().isShutdown() || Stage.MIGRATION.executor().isTerminated())
                {
                    logger.info("Skipped pulling schema from other nodes: the node is decommissioning.");
                    return FINISHED_FUTURE;
                }
```
before this line - https://github.com/apache/cassandra/blob/cf39d031279620eb5684ad384dff72f7325104cd/src/java/org/apache/cassandra/schema/MigrationCoordinator.java#L164

Reply:
I think the same check needs to be placed at this line as well. https://github.com/apache/cassandra/blob/cf39d031279620eb5684ad384dff72f7325104cd/src/java/org/apache/cassandra/schema/MigrationCoordinator.java#L362

The delayed task can be scheduled when the MIGRATION is active. But when it actually runs, the ES could be shutdown already.

Reply:
True, I just added a new commit here - https://github.com/ekaterinadimitrova2/cassandra/commit/da478178dc7ab3d19c8db2f22aca8521d1548207

Reply:
The commit does not seem to have address the issue. Please see the inline comment of the commit.

Reply:
New one as we talked - https://github.com/ekaterinadimitrova2/cassandra/pull/97/commits/bfe0283cd841a86184a8e2ac91959291b4ac4e55

Reply:
and this https://github.com/ekaterinadimitrova2/cassandra/commit/fd9a7fda5ec573dd16374b7b42f566002a0f2843 :D